Source/Core/VideoCommon/BPFunctions.cpp

```diff
@@ -48,24 +49,95 @@ void SetScissor()
   const int xoff = bpmem.scissorOffset.x * 2;
   const int yoff = bpmem.scissorOffset.y * 2;
 
-  EFBRectangle rc(bpmem.scissorTL.x - xoff, bpmem.scissorTL.y - yoff, bpmem.scissorBR.x - xoff + 1,
-                  bpmem.scissorBR.y - yoff + 1);
+  EFBRectangle native_rc(bpmem.scissorTL.x - xoff, bpmem.scissorTL.y - yoff,
+                         bpmem.scissorBR.x - xoff + 1, bpmem.scissorBR.y - yoff + 1);
+
+  if (native_rc.left < 0)
+    native_rc.left = 0;
+  if (native_rc.top < 0)
+    native_rc.top = 0;
+  if (native_rc.right > EFB_WIDTH)
+    native_rc.right = EFB_WIDTH;
+  if (native_rc.bottom > EFB_HEIGHT)
+    native_rc.bottom = EFB_HEIGHT;
+
+  if (native_rc.left > native_rc.right)
+    native_rc.right = native_rc.left;
+  if (native_rc.top > native_rc.bottom)
+    native_rc.bottom = native_rc.top;
+
+  TargetRectangle target_rc = g_renderer->ConvertEFBRectangle(native_rc);
+  g_renderer->SetScissorRect(target_rc);
+}
+
+void SetViewport()
+{
+  int scissor_x_off = bpmem.scissorOffset.x * 2;
+  int scissor_y_off = bpmem.scissorOffset.y * 2;
+  float x = g_renderer->EFBToScaledXf(xfmem.viewport.xOrig - xfmem.viewport.wd - scissor_x_off);
+  float y = g_renderer->EFBToScaledYf(xfmem.viewport.yOrig + xfmem.viewport.ht - scissor_y_off);
```

Comment:
Do you know why OGL had a different sign for xfmem.viewport.ht? There we had used "EFB_HEIGHT - (yOrig - ht - y_off)"

Reply:
This is now handled in the backend. I had assumed it was to flip the framebuffer, due to the lower-left window coordinates. I'll confirm this when I get a chance.

Reply:
The swap is done in the backend, but I haven't seen this for the sum:
`yOrig - ht - y_off`
vs
  `yOrig + ht - y_off`

Through this might be a bug in master...